lib/onelogin/ruby-saml/response.rb

```diff
@@ -89,101 +126,385 @@ def attributes
-            validation_error(error_message)
-          end
+      # Validates if the provided request_id match the inResponseTo value.
+      # If fails, the error is added to the errors array
+      # @param request_id [String|nil] request_id The ID of the AuthNRequest sent by this SP to the IdP (if was sent any)
+      # @param soft [Boolean] soft Enable or Disable the soft mode (In order to raise exceptions when the response is invalid or not)
+      # @return [Boolean] True if there is no request_id or it match, otherwise False if soft=True
+      # @raise [ValidationError] if soft == false and validation fails
+      #
+      def validate_in_response_to(request_id = nil, soft = true)
+        return true if request_id.nil? || in_response_to.nil? || in_response_to.empty?
+
+        unless request_id != in_response_to
+          return true
         end
+
+        error_msg = "The InResponseTo of the Response: #{in_response_to}, does not match the ID of the AuthNRequest sent by the SP: #{request_id}"
+        append_error(soft, error_msg)
       end
 
-      def validate_response_state(soft = true)
-        if response.empty?
-          return soft ? false : validation_error("Blank response")
+      # Validates that there are not EncryptedAttribute (not supported)
+      # If fails, the error is added to the errors array
+      # @param soft [Boolean] soft Enable or Disable the soft mode (In order to raise exceptions when the response is invalid or not)
+      # @return [Boolean] True if there are no EncryptedAttribute elements, otherwise False if soft=True
+      # @raise [ValidationError] if soft == false and validation fails
+      #
+      def validate_no_encrypted_attributes(soft = true)
+        nodes = REXML::XPath.match(
+          document,
+          "/p:Response/a:Assertion/a:AttributeStatement/a:EncryptedAttribute",
+          { "p" => PROTOCOL, "a" => ASSERTION }
+        )
+        if nodes && nodes.length > 0
+          @errors << "There is an EncryptedAttribute in the Response and this SP not support them"
+          return soft ? false : validation_error("There is an EncryptedAttribute in the Response and this SP not support them")
         end
+        true
+      end
 
-        if settings.nil?
-          return soft ? false : validation_error("No settings on response")
+      # Validates the Destination, (if the SAML Response is received where expected)
+      # If fails, the error is added to the errors array
+      # @param soft [Boolean] soft Enable or Disable the soft mode (In order to raise exceptions when the response is invalid or not)
+      # @return [Boolean] True if the destination is valid, otherwise False if soft=True
+      # @raise [ValidationError] if soft == false and validation fails
+      #
+      def validate_destination(soft = true)
+        return true if destination.nil? || destination.empty? || settings.assertion_consumer_service_url.nil? || settings.assertion_consumer_service_url.empty?
```

Comment:
Spec says that destination must be validated if it comes in the assertion, so validation should fail if settings.assertion_consumer_service_url is nil but the assertion contained a non-nil destination.